src/foam/box/CapabilityInterceptBox.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.box',
+  name: "CapabilityInterceptReplyDecoratorBox",
+  extends: 'foam.box.ProxyBox',
+
+  documentation: `
+    This box decorates replyBox of the message with CapabilityInterceptBox
+    before sending the message to its delegate.
+  `,
+
+  methods: [
+    {
+      name: 'send',
+      javaCode: `
+        foam.core.X x = getX();
+
+        msg.getAttributes().put("replyBox",
+          new CapabilityInterceptBox.Builder(x)
+            .setDelegate((Box) msg.getAttributes().get("replyBox"))
+            .build());
+        getDelegate().send(msg);
+      `
+    }
+  ]
+});
+
+foam.CLASS({
+  package: 'foam.box',
+  name: 'CapabilityInterceptBox',
+  extends: 'foam.box.ProxyBox',
+
+  javaImports: [
+    'foam.box.RPCErrorMessage',
+    'foam.box.CapabilityRequiredRemoteException',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.nanos.auth.AuthorizationException',
+    'foam.nanos.auth.User',
+    'foam.nanos.crunch.Capability',
+    'java.util.ArrayList',
+    'java.util.List',
+    'static foam.mlang.MLang.*'
+  ],
+
+  methods: [
+    {
+      name: 'send',
+      javaCode: `
+        foam.core.X x = getX();
+
+        FindCapabilityIntercept: {
+          // Conditions to be met if this is an AuthorizationException reply
+          if ( ! (msg.getObject() instanceof RPCErrorMessage) )
+            break FindCapabilityIntercept;
+          RPCErrorMessage rpcErrMsg = (RPCErrorMessage) msg.getObject();
+          if ( ! (rpcErrMsg.getData() instanceof RemoteException) )
+            break FindCapabilityIntercept;
+          RemoteException remoteEx = (RemoteException) rpcErrMsg.getData();
+          if ( ! (remoteEx.getServerObject() instanceof AuthorizationException) )
+            break FindCapabilityIntercept;
+          AuthorizationException authorEx =
+            (AuthorizationException) remoteEx.getServerObject();
+          String permission = authorEx.getPermission();
+          if ( permission == null || "".equals(permission) )
+            break FindCapabilityIntercept;
+          
+          // Remove permission name from AuthorizationException
+          // (it will still be sent if no capabilities were found)
+          authorEx.clearPermission();
+
+          // Get services needed to detect capability intercepts
+          User user = (User) x.get("user");
+          DAO capabilityDAO = (DAO) x.get("capabilityDAO");
+
+          List<String> capabilityOptions = new ArrayList<String>();
+          {
```

Comment:
What is this for? Is this a syntax error or some language feature I'm not aware of?

Reply:
That's a block statement - it follows the rules of a block without having a loop or condition, so it's equivalent to `if (true) { ... }` but useful to reduce the scope of variables.

I like to use them when I'm creating variables that are only used for the purpose of populating another variable, (in this case `capabilities` is only used to populate `capabilityOptions`) but I admit it's not very commonly used.